### DISCUSSION TITLE
Customize auth/custom endpoints

### DISCUSSION THREAD

**Discussion by pnmcosta** [pocketbase/pocketbase#2412] [Q&A]
I'm building a Shopify React front end App with a Pocketbase back end.

I don't want to bug you much with the details about Shopify embedded app's auth, but it has a [oauth2 process](https://shopify.dev/docs/apps/auth/oauth) (with custom URL per shop installing the app) and includes it's own session token [session token](https://shopify.dev/docs/apps/auth/oauth/session-tokens/getting-started) with all fetches to a back end.

Because I wasn't sure I would be able to tap into PB's auth process to:

- Modify the claim address it's redirected to replace it with the shop's oauth (it's not a static url like other providers).
- Append a middleware/api rule to the default rules that would validate and exchange tokens

I decided to have custom endpoints that handle the custom oauth routes and would modify the collections using `app.Dao()`, but now realizing that events are only raised on the default request routes, I'm not sure proceeding this way is the most optimal, as I was going to hook onto records crud's to process async tasks.

In your opinion, would it be possible to tap the oauth process of one of the spare provider? And would I be able to inject a custom middleware/API Rule that would exchange a session token to a auth collection's one?

I can provide more details if you require.

Thank you,
P.

**Comment by pnmcosta** [comment]
For point one, would I be able to replace `{{shopdomain}}` with any of these [record auth events](https://pocketbase.io/docs/event-hooks/#record-auth-api-hooks) ?

![image](https://user-images.githubusercontent.com/550684/235952973-e6915b18-6f14-46f0-bbe1-7715d1ec48b3.png)

**Comment by ganigeorgiev** [comment]
I'm sorry, I'm not sure that I understand the questions and what exactly you are trying to do.

Could you provide a pseudo-code with the expected workflow?

If you want to modify the app settings programmatically, including the OAuth2 providers, you can do that by updating `app.Settings()`, for example:

```js
// this is without data validations
app.Settings().OIDCAuth.AuthUrl = "something"
app.Dao().SaveSettings(app.Settings())

// if you want data validations, you can use forms.SettingsUpsert
// https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.15.2/forms#SettingsUpsert
form := forms.NewSettingsUpsert(app)
form.OIDCAuth.AuthUrl = "something..."
form.Submit()
```

**Reply by pnmcosta** [reply]
Thanks that's good to know, but I would need to do that in runtime and I believe your solution is to update on boot, correct?

Think of shop as the user, the domain is sent on an initial install request querystring, sequence mermaid below that succinctly explains it as well as the authenticated flow on the front end (2 point)

```mermaid
sequenceDiagram
    Shopify->>BE: be.app/install?shop=shop.url.com
    Note over BE,Shopify: How to runtime redirect?
    BE->>Shopify: shop.url.com/admin/oauth/authorize
    Shopify->>BE: be.app/oauth/callback
    BE->>BE: persist account/access token
    BE->>FE: fe.app/index.html
    
    Note over FE,BE: How to validate/exchange sessionToken?
    FE->>BE: be.app/api/* (fetch w/ sessionToken)
    alt is valid request/token
    BE->>FE: response
    else
    BE->>FE: unauthorized
    end
   ```

**Reply by ganigeorgiev** [reply]
Sorry I still don't think that I understand what do you mean by _"How to runtime redirect?"_ and _"How to validate/exchange sessionToken?"_. It seems that my lack of familiarity with Shopify is preventing me to understand what is the end goal and I'm not sure that I can help.

>  but I would need to do that in runtime and I believe your solution is to update on boot, correct?

You can update the app settings provider while the application is running but it can hold only one value at a time.

Alternatively, you can also configure the OAuth2 provider manually and perform the necessary actions using the [`auth` subpackage](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.15.2/tools/auth), something like:

```go
provider := auth.NewOIDCProvider()
provider.SetAuthUrl("...")
provider.SetUserApiUrl("...")
provider.SetRedirectUrl("...")
token, err := provider.FetchToken(
	"...code...",
	// add any additional code options (eg. for pkce)
	// oauth2.SetAuthURLParam("code_verifier", "...code_verifier..."),
)
```

**Reply by pnmcosta** [reply]
Essentially the oauth process install the app, by granting us a shop's offline access token. The sessionToken is a JWT that identifies a shop's user of the embedded app in the shop's admin. If valid, we use the shop domain to ensure there's a valid shop offline access token already created (oauth process) otherwise we redirect to it.

Got it, cool, would you be open to me contributing with a specific `auth.Shopify` provider in a PR? I can also contribute with the endpoint to the oauth2 package, similar to `AWSCognito` on https://github.com/golang/oauth2/blob/master/endpoints/endpoints.go#L250. It would be a much clearer solution IMO.

I already have a draft middleware to validate the session token as per [docs](https://shopify.dev/docs/apps/auth/oauth/session-tokens/getting-started#obtain-and-verify-session-details). 

Could you clarify if building the custom provider I should be able to ensure that `@request.auth.* - the current authenticated model (eg. @request.auth.id != '')` is validated that way? Or what would be the best approach to support/exchange sessionToken's with PB tokens?

**Reply by ganigeorgiev** [reply]
> Got it, cool, would you be open to me contributing with a specific auth.Shopify provider in a PR?

Sure, feel free to file a PR and I'll review it (although I'm not sure what do you refer by _similar to `AWSCognito`_).

> Could you clarify if building the custom provider I should be able to ensure that @request.auth.* - the current authenticated model (eg. @request.auth.id != '') is validated that way?

Sorry, it seems that I'm again failing to understand what you are trying to explain to me. Initializing manually the OAuth2 provider as shown above (aka. `auth.NewOIDCProvider()`) has nothing to do with the `@request.auth.*` collection filter checks.
`@request.auth.*` checks the fields of the current auth record that is in the current http echo request context.

> Or what would be the best approach to support/exchange sessionToken's with PB tokens?

Again, I don't understand. If you want to generate manually an auth token for a specific record, you can use the [`tokens` subpackage](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.15.2/tokens):

```go
token, err := tokens.NewRecordAuthToken(app, authRecord)
```

Alternatively, if you have an auth record and want to return an auth response from a custom route, you can also return directly `apis.RecordAuthResponse(app, authRecord)`. This will automatically generate the token and write a `{"token": "...", "record": {...}}` json response (_supporting also the `?expand` record enriching_).

**Reply by pnmcosta** [reply]
Ok, so what you're proposing is a custom endpoint that exchanges a sessionToken for a auth record token to authenticate the SDK. I would have to be concerned about expirations, at least keep them in sync, would you have any other concerns?

Apologies for the frustration, I may have mixed some wires, but much obliged for all your help.

**Comment by pnmcosta** [comment]
Ended up not writing a provider, felt the changes required to be able to capture params from the echo context would be to great. It's a pity cause I was hoping I could use the external auth to store some of the shopify auth details like the access token instead of custom record fields, I'd still be happy to contribute, but I feel I would need more of your input, and in the end it might not even be towards your goals for PB.

I am using a custom endpoint to do the session token to pb token exchange for the auth record, but I'm planning to write the FE's SDK wrapper to make use of the localstorage and thus as long as the PB token expiry is longer than the session token we should be fine.

Does the SDK react to deleted auth records? Wondering as it would be neat to kick the user of the FE if that happens.

**Reply by ganigeorgiev** [reply]
> Does the SDK react to deleted auth records?

Yes, if you subscribe to the `users` auth collection you should be able to receive the delete event as long as the client satisfies the List API rule (_if the subscription is for the wildcard topic, `*`_) or the View API rule (_if the subscription is for a single record id_) of the related collection.
